plugins/experimental/fastcgi/src/ats_fcgi_client.cc

```diff
@@ -0,0 +1,646 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ats_fcgi_client.h"
+#include "ats_fastcgi.h"
+#include "fcgi_protocol.h"
+#include "ts/ink_defs.h"
+#include <iostream>
+#include <sstream>
+#include <string>
+#include <locale>
+#include <ts/ts.h>
+#include <map>
+#include <iterator>
+#include "utils_internal.h"
+#include <atscppapi/Headers.h>
+#include <atscppapi/utils.h>
+
+using namespace atscppapi;
+using namespace ats_plugin;
+using namespace std;
+
+struct ats_plugin::FCGIClientState {
+  FCGI_BeginRequest *request;
+  FCGI_Header *header, *postHeader;
+  unsigned char *buff, *pBuffInc;
+  FCGIRecordList *records = nullptr;
+  TSHttpTxn txn_;
+  map<string, string> requestHeaders;
+  int request_id_;
+
+  FCGIClientState()
+    : request(nullptr), header(nullptr), postHeader(nullptr), buff(nullptr), pBuffInc(nullptr), records(nullptr), request_id_(0){};
+
+  ~FCGIClientState()
+  {
+    request_id_ = 0;
+    free(request);
+    free(header);
+    TSfree(postHeader);
+    TSfree(buff);
+    TSfree(records);
+  };
+};
+// input to the constructor will be either unique transaction id or int type
+// requestId
+FCGIClientRequest::FCGIClientRequest(int request_id, TSHttpTxn txn)
+{
+  first_chunk            = true;
+  state_                 = new FCGIClientState();
+  _headerRecord          = nullptr;
+  state_->txn_           = txn;
+  state_->request_id_    = request_id;
+  state_->requestHeaders = GenerateFcgiRequestHeaders();
+  // TODO Call printFCGIRequestHeaders() to printFCGIHeaders
+  // printFCGIRequestHeaders();
+  string str("POST"), value;
+  if (str.compare(state_->requestHeaders["REQUEST_METHOD"]) == 0) {
+    Transaction &transaction = utils::internal::getTransaction(state_->txn_);
+    Headers &h               = transaction.getClientRequest().getHeaders();
+
+    if (h.isInitialized()) {
+      string key("Content-Length");
```

Comment:
What if the POST is chunked?

Reply:
Hmm, searching for 'fastcgi chunked post data' makes it look like it is at least common for FastCGI implementations not to take that well. (top result: https://www.jeffgeerling.com/blog/2017/apache-fastcgi-proxyfcgi-and-empty-post-bodies-chunked-transfer)
Having said that, maybe we should explicitly handle the case and respond with an error when someone tries to POST chunked data, and add a TODO in the code to support it?

Reply:
Testing it out:

```
oschaaf@ats-fastcgi:~/weamp/newats⟫ curl -v -i -d @"test" -H "Transfer-Encoding: chunked"  127.0.0.1/test.aspx
Warning: Couldn't read data from file "test", this makes an empty POST.
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> POST /test.aspx HTTP/1.1
> Host: 127.0.0.1
> User-Agent: curl/7.47.0
> Accept: */*
> Transfer-Encoding: chunked
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 5 out of 0 bytes
< HTTP/1.1 500 Server Error
HTTP/1.1 500 Server Error
< Age: 0
Age: 0
< Date: Thu, 05 Jul 2018 21:17:11 GMT
Date: Thu, 05 Jul 2018 21:17:11 GMT
< Transfer-Encoding: chunked
Transfer-Encoding: chunked
< Connection: keep-alive
Connection: keep-alive
< Server: ATS/9.0.0
Server: ATS/9.0.0

< 
* Connection #0 to host 127.0.0.1 left intact
Content length parameter not an integer.
```

Reply:
(sorry, bad example, but also in the case where I actually do send an *existing*, handling is the same)